File: lib/opal/cli_runners/chrome.rb
# frozen_string_literal: true

require 'shellwords'
require 'socket'
require 'timeout'
require 'tmpdir'
require 'rbconfig'
require 'opal/os'

module Opal
  module CliRunners
    class Chrome
      SCRIPT_PATH = File.expand_path('cdp_interface.rb', __dir__).freeze

      DEFAULT_CDP_HOST = 'localhost'
      DEFAULT_CDP_PORT = 9222

      def self.call(data)
        runner = new(data)
        runner.run
      end

      def initialize(data)
        argv = data[:argv]
        if argv && argv.any?
          warn "warning: ARGV is not supported by the Chrome runner #{argv.inspect}"
        end

        options  = data[:options]
        @output  = options.fetch(:output, $stdout)
        @builder = data[:builder].call
      end

      attr_reader :output, :exit_status, :builder

      def run
        mktmpdir do |dir|
          with_chrome_server do
            prepare_files_in(dir)

            env = {
              'OPAL_CDP_HOST' => chrome_host,
              'OPAL_CDP_PORT' => chrome_port.to_s,
              'NODE_PATH' => File.join(__dir__, 'node_modules'),
              'OPAL_CDP_EXT' => builder.output_extension
            }

            cmd = [
              RbConfig.ruby,
              "#{__dir__}/../../../exe/opal",
              '--no-exit',
              '-I', __dir__,
              '-r', 'source-map-support-node',
              SCRIPT_PATH,
              dir
            ]

            Kernel.exec(env, *cmd)
          end
        end
      end

      private

      def prepare_files_in(dir)
        js = builder.to_s
        map = builder.source_map.to_json
        stack = File.binread("#{__dir__}/source-map-support-browser.js")

        ext = builder.output_extension
        module_type = ' type="module"' if builder.esm?

        # Some maps may contain `</script>` fragment (eg. in strings) which would close our
        # `<script>` tag prematurely. For this case, we need to escape the `</script>` tag.
        map_json = map.to_json.gsub(/(<\/scr)(ipt>)/i, '\1"+"\2')

        # Chrome can't handle huge data passed to `addScriptToEvaluateOnLoad`
        # https://groups.google.com/a/chromium.org/forum/#!topic/chromium-discuss/U5qyeX_ydBo
        # The only way is to create temporary files and pass them to chrome.
        File.binwrite("#{dir}/index.#{ext}", js)
        File.binwrite("#{dir}/index.map", map)
        File.binwrite("#{dir}/source-map-support.js", stack)
        File.binwrite("#{dir}/index.html", <<~HTML)
          <html><head>
            <meta charset='utf-8'>
            <link rel="shortcut icon" href="data:image/x-icon;," type="image/x-icon">
            <script src='./source-map-support.js'></script>
            <script>
            window.OPAL_EXIT_CODE = "noexit"
            sourceMapSupport.install({
              retrieveSourceMap: function(path) {
                return path.endsWith('/index.#{ext}') ? {
                  url: './index.map', map: #{map_json}
                } : null;
              }
            });
            </script>
          </head><body>
            <script src='./index.#{ext}'#{module_type}></script>
          </body></html>
        HTML
      end

      def chrome_host
        ENV['CHROME_HOST'] || ENV['OPAL_CDP_HOST'] || DEFAULT_CDP_HOST
      end

      def chrome_port
        ENV['CHROME_PORT'] || ENV['OPAL_CDP_PORT'] || DEFAULT_CDP_PORT
      end

      def with_chrome_server
        if chrome_server_running?
          yield
        else
          run_chrome_server { yield }
        end
      end

      def run_chrome_server
        raise 'Chrome server can be started only on localhost' if chrome_host != DEFAULT_CDP_HOST

        profile = mktmpprofile

        # Disable web security with "--disable-web-security" flag to be able to do XMLHttpRequest (see test_openuri.rb)
        # For other options see https://github.com/puppeteer/puppeteer/blob/main/packages/puppeteer-core/src/node/ChromeLauncher.ts
        chrome_server_cmd = %{#{OS.shellescape(chrome_executable)} \
          --allow-pre-commit-input \
          --disable-background-networking \
          --disable-background-timer-throttling \
          --disable-backgrounding-occluded-windows \
          --disable-breakpad \
          --disable-client-side-phishing-detection \
          --disable-component-extensions-with-background-pages \
          --disable-crash-reporter \
          --disable-default-apps \
          --disable-dev-shm-usage \
          --disable-extensions \
          --disable-features=Translate,AcceptCHFrame,MediaRouter,OptimizationHints \
          --disable-hang-monitor \
          --disable-infobars \
          --disable-ipc-flooding-protection \
          --disable-popup-blocking \
          --disable-prompt-on-repost \
          --disable-renderer-backgrounding \
          --disable-search-engine-choice-screen \
          --disable-sync \
          --disable-web-security \
          --enable-automation \
          --enable-blink-features=IdleDetection \
          --enable-features=PdfOopif \
          --export-tagged-pdf \
          --force-color-profile=srgb \
          --generate-pdf-document-outline \
          --headless \
          --hide-scrollbars \
          --metrics-recording-only \
          --mute-audio \
          --no-first-run \
          --password-store=basic \
          --use-mock-keychain \
          --user-data-dir=#{profile} \
          --remote-debugging-port=#{chrome_port} \
          #{ENV['CHROME_OPTS']}}

        chrome_pid = Process.spawn(chrome_server_cmd, in: OS.dev_null, out: OS.dev_null, err: OS.dev_null)

        Timeout.timeout(30) do
          loop do
            break if chrome_server_running?
            sleep 0.5
          end
        end

        yield
      rescue Timeout::Error
        puts 'Failed to start chrome server'
        puts 'Make sure that you have it installed and that its version is > 59'
        exit(1)
      ensure
        if OS.windows? && chrome_pid
          Process.kill('KILL', chrome_pid) unless system("taskkill /f /t /pid #{chrome_pid} >NUL 2>NUL")
        elsif chrome_pid
          Process.kill('HUP', chrome_pid)
        end
        FileUtils.rm_rf(profile) if profile
      end

      def chrome_server_running?
        puts "Connecting to #{chrome_host}:#{chrome_port}..."
        TCPSocket.new(chrome_host, chrome_port).close
        true
      rescue Errno::ECONNREFUSED, Errno::EADDRNOTAVAIL
        false
      end

      def chrome_executable
        ENV['GOOGLE_CHROME_BINARY'] ||
          if OS.windows?
            [
              'C:/Program Files/Google/Chrome Dev/Application/chrome.exe',
              'C:/Program Files/Google/Chrome/Application/chrome.exe'
            ].each do |path|
              next unless File.exist? path
              return path
            end
          elsif OS.macos?
            [
              '/Applications/Google Chrome.app/Contents/MacOS/Google Chrome',
              '/Applications/Chromium.app/Contents/MacOS/Chromium'
            ].each do |path|
              next unless File.exist? path
              return path
            end
          else
            %w[
              google-chrome-stable
              chromium
              chromium-freeworld
              chromium-browser
            ].each do |name|
              next unless system('sh', '-c', "command -v #{name.shellescape}", out: '/dev/null')
              return name
            end
            raise 'Cannot find chrome executable'
          end
      end

      def mktmpdir(&block)
        Dir.mktmpdir('chrome-opal-', &block)
      end

      def mktmpprofile
        Dir.mktmpdir('chrome-opal-profile-')
      end
    end
  end
end

File: lib/opal/cli_runners/chrome.js
var CDP = require("chrome-remote-interface");
var fs = require("fs");

// NOTE: the code above doesn't work on Windows
// and it doesn't support interactive stdin input (which is ok for our scenario)
var opal_code = fs.readFileSync("/dev/stdin").toString();

// Chrome can't handle huge data passed to `addScriptToEvaluateOnLoad`
// https://groups.google.com/a/chromium.org/forum/#!topic/chromium-discuss/U5qyeX_ydBo
// The only way is to create temporary files and pass them to chrome.
fs.writeFileSync("/tmp/chrome-opal.js", opal_code);
fs.writeFileSync("/tmp/chrome-opal.html", "" +
  "<html>" +
  "<head>" +
    "<meta charset='utf-8'>" +
  "</head>" +
  "<body>" +
    "<script src='./chrome-opal.js'></script>" +
  "</body>" +
  "</html>"
);

var options = {
  host: process.env.CHROME_HOST || 'localhost',
  port: process.env.CHROME_PORT || 9222
};

CDP(options, function(client) {
  var Page = client.Page,
      Runtime = client.Runtime,
      Console = client.Console;

  Promise.all([
    Console.enable(),
    Page.enable(),
    Runtime.enable(),
  ]).then(function() {
    // This hook catches only the first argument of `console.log`
    // More advanced version Runtime.consoleAPICalled returns all arguments
    // but all of them are not formatted, i.e. by calling
    //   console.log('string', [1, 2, 3], {a: 'b'})
    // it returns the following data to the callback:
    //   {
    //     "type":"log",
    //     "args":[
    //       {
    //         "type":"string",
    //         "value":"string"
    //       },
    //       {
    //         "type":"object",
    //         "subtype":"array",
    //         "className":"Array",
    //         "description":"Array(3)",
    //         "objectId":"{\"injectedScriptId\":11,\"id\":1}",
    //         "preview":{
    //           "type":"object",
    //           "subtype":"array",
    //           "description":"Array(3)",
    //           "overflow":false,
    //           "properties":[
    //             {"name":"0","type":"number","value":"1"},
    //             {"name":"1","type":"number","value":"2"},
    //             {"name":"2","type":"number","value":"3"}
    //           ]
    //         }
    //       },
    //       {
    //         "type":"object",
    //         "className":"Object",
    //         "description":"Object",
    //         "objectId":"{\"injectedScriptId\":11,\"id\":2}",
    //         "preview":{
    //           "type":"object",
    //           "description":"Object",
    //           "overflow":false,
    //           "properties":[
    //             {"name":"a","type":"string","value":"b"}
    //           ]
    //         }
    //       }
    //     ],
    //     // ...
    //   }
    // Supporting this format for complex data structure is challenging, feel free to contribute!
    //
    Console.messageAdded(function(console_message) {
      process.stdout.write(console_message.message.text);
    });

    Runtime.exceptionThrown(function(exception) {
      var exceptionDetails = exception.exceptionDetails,
          properties = exceptionDetails.exception.preview.properties,
          stackTrace = exceptionDetails.stackTrace ? exceptionDetails.stackTrace.callFrames : [],
          name, message, trace = [], i;


      for (i = 0; i < properties.length; i++) {
        var property = properties[i];

        if (property.name == "name") {
          name = property.value;
        }

        if (property.name == "message") {
          message = property.value;
        }
      }

      console.log(name + " : " + message);

      for (i = 0; i < stackTrace.length; i++) {
        var raw = stackTrace[i];

        console.log("    at " + raw.functionName + " (" + raw.url + ":" + raw.lineNumber + ":" + raw.columnNumber + ")")
      }

      process.exit(1);
    })

    Page.loadEventFired(() => {
      Runtime.evaluate({ expression: "window.OPAL_EXIT_CODE" }).then(function(output) {
        client.close();

        if (typeof(output.result) !== "undefined" && output.result.type === "number") {
          process.exit(output.result.value);
        } else {
          process.exit(0);
        }
      })
    });

    Page.navigate({ url: 'file:///tmp/chrome-opal.html' })
  });
});
